Continuously fixing compiling warnings with this issue

### Description of the enhancement or error report
Occasionally we introduce compiling warnings in the debug mode, which can only be seen with `METHOD=dbg make`.

### Rationale for the enhancement or information for reproducing the error
Users can report warnings here. Then moose developers can reproduce the warning with the debug make.

### Identified impact
Typically the warnings are trivial. But sometimes they could be indicating a serious bug.

/Users/wangy2/projects/moose/framework/include/vectorpostprocessors/PointSamplerBase.h:43:16: warning: 'PointSamplerBase::threadJoin' hides overloaded virtual function [-Woverloaded-virtual]

  virtual void threadJoin(const SamplerBase & y);

               ^

/Users/wangy2/projects/moose/framework/include/userobject/GeneralUserObject.h:57:16: note: hidden overloaded virtual function 'GeneralUserObject::threadJoin' declared here: type mismatch at 1st parameter ('const UserObject &' vs 'const SamplerBase &')

  virtual void threadJoin(const UserObject &) /*final*/;

               ^

In file included from /Users/wangy2/projects/moose/framework/src/base/Moose.C:221:

In file included from /Users/wangy2/projects/moose/framework/include/vectorpostprocessors/LineMaterialRealSampler.h:19:

/Users/wangy2/projects/moose/framework/include/vectorpostprocessors/LineMaterialSamplerBase.h:87:16: warning: 'LineMaterialSamplerBase<double>::threadJoin' hides overloaded virtual function [-Woverloaded-virtual]

  virtual void threadJoin(const SamplerBase & sb);

               ^

/Users/wangy2/projects/moose/framework/include/vectorpostprocessors/LineMaterialRealSampler.h:32:10: note: in instantiation of template class 'LineMaterialSamplerBase<double>' requested here

  public LineMaterialSamplerBase<Real>

         ^

/Users/wangy2/projects/moose/framework/include/userobject/GeneralUserObject.h:57:16: note: hidden overloaded virtual function 'GeneralUserObject::threadJoin' declared here: type mismatch at 1st parameter ('const UserObject &' vs 'const SamplerBase &')

  virtual void threadJoin(const UserObject &) /*final*/;

In file included from /Users/wangy2/projects/moose/framework/src/outputs/ICEUpdater.C:18:

/Users/wangy2/projects/moose/framework/contrib/ice_updater/include/Updater.h:432:25: warning: unused parameter 'stream' [-Wunused-parameter]

  Updater(std::istream &stream) { mooseError("Updater requires --enable-cxx11 parameter to update_and_build_libmesh.sh."); }

                        ^

1 warning generated.

--- a/framework/include/vectorpostprocessors/LineMaterialSamplerBase.h
+++ b/framework/include/vectorpostprocessors/LineMaterialSamplerBase.h
@@ -80,13 +80,6 @@ public:
   virtual void finalize();
 
   /**
-   * Thread Join
-   * Calls through to base class's threadJoin()
-   * @param sb SamplerBase object to be joint into this object
-   */
-  virtual void threadJoin(const SamplerBase & sb);
-
-  /**
    * Reduce the material property to a scalar for output
    * @param property The material property
    * @param curr_point The point corresponding to this material property
@@ -197,13 +190,6 @@ void
 LineMaterialSamplerBase<T>::finalize()
 {
   SamplerBase::finalize();
-}
-
-template <typename T>
-void
-LineMaterialSamplerBase<T>::threadJoin(const SamplerBase & sb)
-{
-  SamplerBase::threadJoin(sb);
 }
 
 #endif

--- a/framework/include/vectorpostprocessors/PointSamplerBase.h
+++ b/framework/include/vectorpostprocessors/PointSamplerBase.h
@@ -40,8 +40,6 @@ public:
   virtual void execute();
   virtual void finalize();
 
-  virtual void threadJoin(const SamplerBase & y);
-
 protected:
   /**
    * Find the local element that contains the point.  This will attempt to use a cached element to speed things up.

--- a/framework/include/vectorpostprocessors/SamplerBase.h
+++ b/framework/include/vectorpostprocessors/SamplerBase.h
@@ -38,8 +38,6 @@ public:
    * @param comm The communicator of the child
    */
   SamplerBase(const InputParameters & parameters, VectorPostprocessor * vpp, const libMesh::Parallel::Communicator & comm);
-  SamplerBase(const std::string & /*name*/, InputParameters parameters, VectorPostprocessor * vpp,
-              const libMesh::Parallel::Communicator & comm); // DEPRECATED CONSTRUCTOR
   virtual ~SamplerBase() {}
 
 protected:

--- a/framework/src/vectorpostprocessors/PointSamplerBase.C
+++ b/framework/src/vectorpostprocessors/PointSamplerBase.C
@@ -125,15 +125,6 @@ PointSamplerBase::finalize()
   SamplerBase::finalize();
 }
 
-void
-PointSamplerBase::threadJoin(const SamplerBase & y)
-{
-  const PointSamplerBase & vpp = static_cast<const PointSamplerBase &>(y);
-
-  SamplerBase::threadJoin(vpp);
-}
-
-
 const Elem *
 PointSamplerBase::getLocalElemContainingPoint(const Point & p, unsigned int /*id*/)
 {

--- a/framework/src/vectorpostprocessors/SamplerBase.C
+++ b/framework/src/vectorpostprocessors/SamplerBase.C
@@ -145,17 +145,3 @@ SamplerBase::threadJoin(const SamplerBase & y)
   for (unsigned int i=0; i<_variable_names.size(); i++)
     _values_tmp[i].insert(_values_tmp[i].end(), y._values_tmp[i].begin(), y._values_tmp[i].end());
 }
-
-// DEPRECATED CONSTRUCTOR
-SamplerBase::SamplerBase(const std::string & /*name*/, InputParameters parameters, VectorPostprocessor * vpp,
-                         const libMesh::Parallel::Communicator & comm) :
-    _sampler_params(parameters),
-    _vpp(vpp),
-    _comm(comm),
-    _sort_by(parameters.get<MooseEnum>("sort_by")),
-    _x(vpp->declareVector("x")),
-    _y(vpp->declareVector("y")),
-    _z(vpp->declareVector("z")),
-    _id(vpp->declareVector("id"))
-{
-}

--- a/test/src/userobjects/ReadDoubleIndex.C
+++ b/test/src/userobjects/ReadDoubleIndex.C
@@ -101,14 +101,14 @@ ReadDoubleIndex::ReadDoubleIndex(const InputParameters & params) :
   // check int
   if (_int_di.size() != 3)
     mooseError("Error reading int_di.");
-  for (int j = 0; j < 3; ++j)
+  for (unsigned int j = 0; j < 3; ++j)
   {
     if (_int_di[j].size() != array_length[j])
       mooseError("Error reading int_di.");
-    for (int i = 0; i < _int_di[j].size(); ++i)
+    for (unsigned int i = 0; i < _int_di[j].size(); ++i)
     {
       int mult = ((j == 1) ? -1 : 1);
-      if (_int_di[j][i] != mult * ((j + 1) * 10 + i + 1))
+      if (_int_di[j][i] != mult * static_cast<int>(((j + 1) * 10 + i + 1)))
         mooseError("Error reading int_di.");
     }
   }
@@ -116,11 +116,11 @@ ReadDoubleIndex::ReadDoubleIndex(const InputParameters & params) :
   // check long
   if (_long_di.size() != 3)
     mooseError("Error reading long_di.");
-  for (long j = 0; j < 3; ++j)
+  for (unsigned int j = 0; j < 3; ++j)
   {
     if (_long_di[j].size() != array_length[j])
       mooseError("Error reading long_di.");
-    for (long i = 0; i < _long_di[j].size(); ++i)
+    for (unsigned int i = 0; i < _long_di[j].size(); ++i)
     {
       long mult = ((j != 1) ? -1 : 1);
       if (_long_di[j][i] != mult * ((j + 1) * 10 + i + 1))
@@ -148,7 +148,7 @@ ReadDoubleIndex::ReadDoubleIndex(const InputParameters & params) :
     if (_bid_di[j].size() != array_length[j])
       mooseError("Error reading bid_di.");
     for (unsigned int i = 0; i < _bid_di[j].size(); ++i)
-      if (_bid_di[j][i] != (j + 2) * 10 + i + 1)
+      if (_bid_di[j][i] != static_cast<short>((j + 2) * 10 + i + 1))
         mooseError("Error reading bid_di.");
   }
 

